test/metabase/data_apps/test_util.clj

```diff
@@ -0,0 +1,79 @@
+(ns metabase.data-apps.test-util
+  "Test utilities for data apps"
+  (:require
+   [metabase.data-apps.models :as data-apps.models]
+   [metabase.test :as mt]
+   [metabase.test.util :as tu]
+   [metabase.util.malli :as mu]))
+
+(defmacro with-data-app-cleanup!
+  "Execute `body`, then delete any *new* data app-related rows created during execution.
+  This includes DataApp, DataAppDefinition, and DataAppRelease models.
+
+  Example:
+    (with-data-app-cleanup
+      (create-data-app-via-api!)
+      (is (= ...)))"
+  [& body]
+  `(tu/with-model-cleanup [:model/DataApp :model/DataAppDefinition :model/DataAppRelease]
+     ~@body))
+
+(defn do-with-data-app!
+  "Create a temporary data app for testing. Optionally creates a definition if :definition is provided."
+  [{:keys [definition] :as data-app} thunk]
+  (mt/with-temp
+    [:model/DataApp {app-id :id :as app} (merge
+                                          {:name "Test Data App"}
+                                          (dissoc data-app :definition))]
+    (let [app-with-definition (if definition
+                                (do
+                                  (data-apps.models/set-latest-definition! app-id {:config definition
+                                                                                   :creator_id (mt/user->id :crowberto)})
+                                  (assoc app :definition (data-apps.models/latest-definition app-id)))
```

Comment:
Won't `latest-definition` just return `definition`?

Reply:
At worst it would be nice to use the return value from `set-latest-definition!` instead of making another query, but even that seems like overkill?